src/cdk-experimental/testing/component-harness.ts

```diff
@@ -8,183 +8,267 @@
 
 import {TestElement} from './test-element';
 
-/** Options that can be specified when querying for an Element. */
-export interface QueryOptions {
-  /**
-   * Whether the found element can be null. If allowNull is set, the searching function will always
-   * try to fetch the element at once. When the element cannot be found, the searching function
-   * should return null if allowNull is set to true, throw an error if allowNull is set to false.
-   * If allowNull is not set, the framework will choose the behaviors that make more sense for each
-   * test type (e.g. for unit test, the framework will make sure the element is not null; otherwise
-   * throw an error); however, the internal behavior is not guaranteed and user should not rely on
-   * it. Note that in most cases, you don't need to care about whether an element is present when
-   * loading the element and don't need to set this parameter unless you do want to check whether
-   * the element is present when calling the searching function. e.g. you want to make sure some
-   * element is not there when loading the element in order to check whether a "ngif" works well.
-   */
-  allowNull?: boolean;
-  /**
-   * If global is set to true, the selector will match any element on the page and is not limited to
-   * the root of the harness. If global is unset or set to false, the selector will only find
-   * elements under the current root.
-   */
-  global?: boolean;
-}
+/** An async function that returns a promise when called. */
+export type AsyncFn<T> = () => Promise<T>;
 
-/** Interface that is used to find elements in the DOM and create harnesses for them. */
-export interface HarnessLocator {
+/**
+ * Interface used to load ComponentHarness objects. This interface is used by test authors to
+ * instantiate `ComponentHarness`es.
+ */
+export interface HarnessLoader {
```

Comment:
Should we turn this into an abstract class instead? That way we could consolidate some of the behavior like in `findRequired` where it's supposed to throw an error.

Reply:
There is actually an abstract class `HarnessEnvironment` that implements many of the methods on `HarnessLoader` and `LocatorFactory`.

I _could_ probably separate them into 3 separate classes instead of 1 class that implements the 2 interfaces. It wound up as one mostly just because of how it evolved from the old implementation.

Reply:
Discussed with @jelbourn and decided to just keep as is for now